[RFC] TA profiling with gprof

jforissier opened the pull request:
Depends on tee-supplicant patch: https://github.com/OP-TEE/optee_client/pull/61.

Example usage:

```
# Build OP-TEE with TA profiling enabled and instrument libutee
$ make -C optee_os ... CFG_TEE_CORE_USER_MEM_DEBUG=0 CFG_TA_GPROF_SUPPORT=y CFG_LIBUTEE_GPROF=y

# Build xtest with profiling enabled
$ make -C optee_test ... CFLAGS_ta_arm64="-O0 -pg"

# Run xtest on the target
root@deb-est:/tmp xtest 4002
<snip>
INFO:    USER-TA: gprof: initializing (buffer size: 201036, text size: 116888)
<snip>
+-----------------------------------------------------
Result of testsuite XTEST_TEE_TEST filtered by "4002":
XTEST_TEE_4002 OK
+-----------------------------------------------------
549 subtests of which 0 failed
1 test case of which 0 failed
713 test cases was skipped
TEE test application done!

# Run gprof on the target or on the host
$ gprof --line cb3e5ba0-adf1-11e0-998b0002a5d5c51b.elf gmon-cb3e5ba0-adf1-11e0-998b0002a5d5c51b.out
```

The gprof output is here: https://gist.github.com/jforissier/91da987925f2e9a7e86a0a84f1f25909

jenswi-linaro commented on lib/libutee/arch/arm/gprof/gprof.c:
```diff
@@ -0,0 +1,342 @@
+ *   gmon/gmon.c
+ *   gmon/mcount.c
+ * Copyright (c)
+ *      The Regents of the University of California.  All rights reserved.
```
I guess we need to license too.

jenswi-linaro replied:
Was the original files under the exact license we have above? If so why not just put it below the Linaro copyright above to avoid future questions.

jforissier replied:
Ah, got it. You're right -- I'll add the full license text.

jenswi-linaro commented on lib/libutee/arch/arm/gprof/gprof.c:
```diff
@@ -0,0 +1,342 @@
+}
+
+/* Adjust PC so that gprof can locate it in the TA ELF file */
+#define ADJUST_PC(pc) \
```
This way of updating parameters to a macro is something I think should be avoided unless the alternative is to inconvenient, in this case it isn't.

jenswi-linaro replied:
On the other hand, if this is imported code we should keep it as it is.

jforissier replied:
This part is mine ;) I'll change it to `pc = adjust_pc(pc);`

jenswi-linaro commented on lib/libutee/arch/arm/gprof/gprof.h:
```diff
@@ -0,0 +1,256 @@
+ * This part adapted from glibc:
+ *   gmon/sys/gmon_out.h
+ *
+ * Copyright (C) 1996-2016 Free Software Foundation, Inc.
```
Is this license really OK here?

jenswi-linaro replied:
Looks OK, I think this should be explained here, because it may look a bit scary otherwise.
@jbech-linaro what do you think?

jbech-linaro replied:
Yes, wouldn't hurt to explain and refer to the section Jerome linked to.  In general everything that isn't BSD Clause-2 needs to be dealt with a bit more carefully. So, this is LGPL3, looking at: http://choosealicense.com/licenses/lgpl-3.0 it says: Disclose Source (Source code must be made available when distributing the software), that is fine for what we have on GitHub, but what about the people (read companies) using OP-TEE? I'm not sure they'd be happy about it. Am I wrong or is this again the grey zone we're touching?

etienne-lms replied:
Disabling `CFG_*_GPROF*` should allow users not fear licensing issues.
Could op-tee be released under gpl-like when features are enabled, and under an alternate 2-clause BSD when those features are disable ?

jenswi-linaro replied:
No, I don't think we should have different features depending on which license we're compiling with. It's begging for problems and conflicts of interest.

Looking at the original `gmon_out.h`, I'd prefer keeping it as a separate file (with required modification to make it useful). Both to avoid mixing of licenses in a single file and also because it will be more clear that it's just a borrowed .h file.

jforissier replied:
@jenswi-linaro keeping two separate files is a good idea (or rather: merging them was a bad idea).

@jbech-linaro I do think it's OK to have `gmon_out.h` with the LGPL license. The section you're citing (i.e.: disclose source) won't apply specifically because of the "header documenting structure layout" exception. I'll mention it clearly.

jbech-linaro replied:
> @jbech-linaro I do think it's OK to have gmon_out.h with the LGPL license. The section you're citing (i.e.: disclose source) won't apply specifically because of the "header documenting structure layout" exception. I'll mention it clearly.

I'm still not 100% convinced about licenses here. Instead of us "thinking" that it is OK, I'd like have someone knowing for sure, saying whether this is OK or not.

Also, we haven't fulfilled neither a) or b) on the link you pasted above pointing to the LGPL license. Can we please wait merge this until we're sure about it?

jbech-linaro replied:
https://www.youtube.com/watch?v=JlIrSMzF8T4 , my understanding is that any company using OP-TEE are not enforced to give away the source for the major part of OP-TEE, but they must give away the LGPL code from OP-TEE. That might not sound too bad, but I can imagine there are companies using OP-TEE that don't want to deal with this at all. I.e, they don't want to spend time on picking subsets of OP-TEE in case someone asks for it.

Having that said, we could poll the known users of OP-TEE, but that is not a very good way of dealing with questions like this.

As usual, I'm unsure whether the license covers compiled code, or both that and the source code available in the branch (other branches cannot impact I hope). I you have experience in this area, please let me know.

Fallback? Put it on a separate branch as of now?

etienne-lms commented:
Disabling CFG_*_GPROF* should allow users not fear licensing issues.
Could op-tee be released under xgpl when features are enabled, and under an alternate 2-clause BSD when those features are disable ?

jforissier commented on lib/libutils/ext/include/compiler.h:
```diff
@@ -53,6 +53,7 @@
 #define __rodata	__section(".rodata")
 #define __rodata_unpaged __section(".rodata.__unpaged")
 #define __early_bss	__section(".early_bss")
+#define __notrace	__attribute__((no_instrument_function))
```
Sounds good, so I'll change to use that instead.
FYI, the linux kernel uses `notrace` with no leading underscores (which is not OK here because `compiler.h` may be used in user applications so underscores are needed to avoid potential conflicts).

jforissier commented:
@etienne-lms the standard C library (glibc) creates a file in the current directory of the application that's been compiled with `-pg` (the file is called `gmon.out`). Having a similar file created by `tee-supplicant` was a natural thing to do in my opinion (and quite simple indeed). I don't think there is any security issue with this because the TA has to be explicitly compiled with profiling enabled (`-pg`), and even the profiling support in libutee/TEE core/tee-supplicant would be disabled by default in production builds I suppose. That's really a development feature.

jforissier commented on core/arch/arm/tee/arch_svc.c:
```diff
@@ -134,6 +135,7 @@ static const struct syscall_entry tee_svc_syscall_table[] = {
 	SYSCALL_ENTRY(syscall_se_channel_transmit),
 	SYSCALL_ENTRY(syscall_se_channel_close),
 	SYSCALL_ENTRY(syscall_cache_operation),
+	SYSCALL_ENTRY(syscall_gprof_send),
```
What's the other option(s)? A static TA? Would be OK if it were as simple to implement and invoke from `libutee` as a syscall. Is it?

jforissier commented:
Rebased on top of master due to the re-organization of `optee_msg.h` / `optee_msg_fs.h` (deleted) / `optee_msg_supplicant.h`. In the meantimg I have squashed all the changes that address the review comments until now into the main commit ("Add support...").
Next I'll replace the syscall with a static TA.

jforissier commented:
On Thu, Oct 27, 2016 at 2:50 PM, Joakim Bech <notifications@github.com>
wrote:

> *@jbech-linaro* commented on this pull request.
> ------------------------------
>
> In lib/libutee/arch/arm/gprof/gprof.h
> <https://github.com/OP-TEE/optee_os/pull/1128>:
>
> > +
> +/*
> + * Possible states of profiling.
> + */
> +#define	GMON_PROF_ON	0
> +#define	GMON_PROF_BUSY	1
> +#define	GMON_PROF_ERROR	2
> +#define	GMON_PROF_OFF	3
> +
> +/*
> + * gmon.out file format
> + *
> + * This part adapted from glibc:
> + *   gmon/sys/gmon_out.h
> + *
> + * Copyright (C) 1996-2016 Free Software Foundation, Inc.
>
> @jbech-linaro <https://github.com/jbech-linaro> I do think it's OK to
> have gmon_out.h with the LGPL license. The section you're citing (i.e.:
> disclose source) won't apply specifically because of the "header
> documenting structure layout" exception. I'll mention it clearly.
>
> I'm still not 100% convinced about licenses here. Instead of us "thinking"
> that it is OK, I'd like have someone knowing for sure, saying whether this
> is OK or not.
>
Sure. Good luck finding someone who will sign in blood ;)

> Also, we haven't fulfilled neither a) or b) on the link you pasted above
> pointing to the LGPL license. Can we please wait merge this until we're
> sure about it?
>
You have to do a) and b) when the header file you're using is *not* limited
to data structure layout. Our copy is just data structure layout. So we
don't have to do a) nor b).

—
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/OP-TEE/optee_os/pull/1128>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AH53w-pwl-276jXpRBz5Em5_HvwPctXjks5q4J38gaJpZM4KdUfy>
> .
>

jforissier commented:
New example of gprof output, with time statistics thanks to the PC sampling support: https://gist.github.com/jforissier/4f26dc82d8e844cb795ea18782cccfc0
The TA is a modified version of xtest, running two functions `wait1(2000);` and `wait2(3000);`. Each function waits for the specified amount of time (ms) by looping and periodically checking the value of CNTVCT. This is clearly visible in the profile. The absolute time values seem a bit off, probably due to the hard-coded value I used for the sampling frequency.

jforissier commented:
Updated. The full `gprof` output for `xtest 1006` on QEMU is here: https://gist.github.com/jforissier/e8cfd623c1824df4538acdb4caa4b45e.

jforissier commented:
Closing this RFC PR since I have open a new, consolidated one.
